### PR TITLE
Improve Reader View performance, rewrite BrowserPanel

### LINKED ISSUE #200: Reader View is too slow
Reader View works but could be faster with some optimization.
For example, if a page is already downloaded, reader view could be generated from the file itself rather than from the browser's document object.
Another example is that changing reader view styles causes an entire reload instead of just changing the stylesheet

### PR DESCRIPTION
**Pull request by rlrh**
Resolves #200 
#### Summary of changes:
- Rewrite BrowserPanel - should be much more reliable
- Optimize Reader View performance
- Display browser loading progress bar

### CODE REVIEW

File: src/main/java/seedu/address/ui/BrowserPanel.java [hunk 1]
```diff
@@ -48,45 +47,29 @@
...
-                loadPage(webEngine.getLocation());
-            } else if (currentLocationUrl != null) {
-                loadPage(currentLocationUrl.toString());
-            }
-        });
-
-        // Respond to browser state events.
+        // React to browser loading state changes.
         webEngine.getLoadWorker().stateProperty().addListener((observable, oldState, newState) -> {
             switch (newState) {
```
Comment (thomastanck) [inline src/main/java/seedu/address/ui/BrowserPanel.java]: Can probably extract this out to handleBrowserLoadChange(Worker.State newState)

File: src/main/java/seedu/address/ui/BrowserPanel.java [hunk 2]
```diff
@@ -103,149 +86,277 @@ public BrowserPanel(ObservableValue<Entry> selectedEntry,
...
-
     }
 
+    //=========== Loading state handlers ===============================================================================
+
     /**
-     * Displays loading message.
+     * Handler for running state.
      */
     private void handleRunning() {
```
Comment (thomastanck) [inline src/main/java/seedu/address/ui/BrowserPanel.java:125]: ... and then inline handleRunning, handleSucceeded, handleFailed.

File: src/main/java/seedu/address/ui/BrowserPanel.java [hunk 3]
```diff
@@ -103,149 +86,277 @@ public BrowserPanel(ObservableValue<Entry> selectedEntry,
...
+                setStyleSheet(newViewMode.getReaderViewStyle().getStylesheetLocation());
+            } else if (newViewMode.hasReaderViewType() && currentlyInOfflinePage()) {
+                getArticle
+                        .apply(lastEntryUrl)
+                        .ifPresentOrElse(this::loadReaderOfLastEntry, this::loadReaderOfLastEntry);
+            } else if (newViewMode.hasReaderViewType() && currentlyInOnlinePage()) {
+                loadReaderOfLastUrl();
+            } else if (newViewMode.hasBrowserViewType() && !currentlyInExternalPage()) {
+                loadPageOfLastUrl();
+            }
```
Comment (thomastanck) [inline src/main/java/seedu/address/ui/BrowserPanel.java]: What happens if none of the conditionals pass? (is it even possible?)

**Review by thomastanck — APPROVED** [review]

File: src/main/java/seedu/address/ui/BrowserPanel.java (hunk 3 above)
Comment (rlrh) [reply src/main/java/seedu/address/ui/BrowserPanel.java]: It's definitely possible, like if you're in an online page and you change to browser view - in this case no action needs to be taken.

Comment (thomastanck) [reply src/main/java/seedu/address/ui/BrowserPanel.java]: Maybe add a comment describing when none of the conditionals pass and explain why no action needs to be taken?

Comment (rlrh) [reply src/main/java/seedu/address/ui/BrowserPanel.java]: Resolved in de41f6e

File: src/main/java/seedu/address/ui/BrowserPanel.java (hunk 1 above)
Comment (rlrh) [reply src/main/java/seedu/address/ui/BrowserPanel.java]: Done in de41f6e

File: src/main/java/seedu/address/ui/BrowserPanel.java (hunk 2 above)
Comment (rlrh) [reply src/main/java/seedu/address/ui/BrowserPanel.java:125]: I decided not to inline as bringing them together would make the code for that method quite long.

**Pull request merged**